Add a validator for documentation package names and base urls.

Numerlor opened the pull request:
https://github.com/python-discord/bot/pull/1014 will introduce a restriction on the package names to `a-z`, `0-9` and `_` because they will be used directly for stats and will enforce base urls to end with slashes. This PR implements the restrictions on the site side.

Numerlor committed:
Add a validator for package names.

Package names are used for stats in the bot and are restricted to
the a-z_ char set, a validator is added to accommodate this restriction
at the site admin side.

Numerlor committed:
Add ascii digits to the validator.

Some packages can contain them and are good for stats.

Numerlor committed:
Move package name validator definition.

The move prevents it going through the line limit
and deeper nesting of parentheses from splitting up the string.

Numerlor committed:
Fix package name validator definition.

The validators kwarg expects an iterable of validators,
while a validator directly was being supplied.

Numerlor committed:
Create migration for doc package name validator.

lemonsaurus reviewed (approved):
This looks good to me.

MarkKoz reviewed (changes requested):
You should document in the viewset that a 400 will be returned by the POST due to the regex validator failing.

Numerlor committed:
Merge branch 'main' into doc-validator

Numerlor committed:
Ensure the base url ends with a slash

URLs without a trailing slash won't get properly joined by the bot
Adds compatibility with python-discord/bot@bc25bfdf42cdaaba924a7ad6de1dc06a9b381285's changes

The styling of how the regex decorator is constructed was changed to be consisted with the
function validator

Numerlor committed:
Update tests to use trailing slashes on valid urls

MarkKoz committed:
Merge branch 'main' into doc-validator
